File: server/app/api/projects.py
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from ..db.session import get_db
from ..schemas.project import (
    ProjectCreate,
    ProjectRead,
    ProjectPatch,
    ProjectStats,
    ProjectBundle,
    ProjectExportRequest,
    ProjectImportResult,
)
from ..services import projects as svc

router = APIRouter(prefix="/api/projects")


def _wrap(data):
    return {"ok": True, "data": data}


# v2.0 T6 — 列表端点用 success/data/meta 三段式信封；其它端点保留旧的 ok/data 信封
# 以避免改动前端契约。
def _read(p) -> dict:
    return ProjectRead.model_validate(p).model_dump(mode="json")


class ProjectCopyIn(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)


@router.post("", status_code=201)
def create(payload: ProjectCreate, db: Session = Depends(get_db)):
    p = svc.create(db, payload)
    return _wrap(_read(p))


@router.get("")
def list_all(
    q: str | None = Query(None, max_length=100),
    city: str | None = Query(None, max_length=20),
    industry: str | None = Query(None, max_length=40),
    phase: str | None = Query(None, pattern="^(budget|bidding|planning|change|settled)$"),
    mode: str | None = Query(None, pattern="^(forward|reverse)$"),
    sort: str = Query("created_at", pattern="^(created_at|updated_at|name|target_cost)$"),
    order: str = Query("desc", pattern="^(asc|desc)$"),
    page: int = Query(1, ge=1),
    size: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
) -> dict:
    rows, total = svc.list_with_query(
        db,
        q=q,
        city=city,
        industry=industry,
        phase=phase,
        mode=mode,
        sort=sort,
        order=order,
        page=page,
        size=size,
    )
    return {
        "success": True,
        "data": [_read(p) for p in rows],
        "error": None,
        "meta": {"total": total, "page": page, "size": size},
    }


@router.get("/stats", response_model=ProjectStats)
def get_project_stats(month: str | None = None, db: Session = Depends(get_db)):
    return svc.get_stats(db, month=month)


@router.post("/export")
def export_projects(payload: ProjectExportRequest, db: Session = Depends(get_db)) -> dict:
    bundle = svc.export_projects(db, payload.ids)
    return {"success": True, "data": bundle, "error": None}


@router.post("/import")
def import_projects(payload: dict, db: Session = Depends(get_db)) -> dict:
    from pydantic import ValidationError
    try:
        bundle = ProjectBundle.model_validate(payload)
    except ValidationError as e:
        raise HTTPException(
            400,
            detail={"error": {
                "code": "INVALID_BUNDLE",
                "message": "导入数据格式非法",
                "problem": str(e.errors()[0].get("msg", "格式校验失败")),
            }},
        )
    n, ids = svc.import_bundle(db, bundle)
    result = ProjectImportResult(imported=n, project_ids=ids)
    return {"success": True, "data": result.model_dump(mode="json"), "error": None}


@router.get("/{project_id}")
def get_one(project_id: str, db: Session = Depends(get_db)):
    p = svc.get(db, project_id)
    if not p:
        raise HTTPException(404, detail={"error": {"code": "NOT_FOUND"}})
    return _wrap(_read(p))


@router.patch("/{project_id}")
def patch_one(project_id: str, payload: ProjectPatch, db: Session = Depends(get_db)):
    p = svc.patch(db, project_id, payload)
    if not p:
        raise HTTPException(404, detail={"error": {"code": "NOT_FOUND"}})
    return _wrap(_read(p))


@router.delete("/{project_id}")
def delete_one(project_id: str, db: Session = Depends(get_db)):
    ok = svc.delete(db, project_id)
    if not ok:
        raise HTTPException(404, detail={"error": {"code": "NOT_FOUND"}})
    return _wrap({"deleted": project_id})


@router.post("/{project_id}/copy", status_code=201)
def copy(project_id: str, payload: ProjectCopyIn, db: Session = Depends(get_db)) -> dict:
    try:
        new = svc.copy_project(db, project_id, payload.name)
    except ValueError as e:
        if "NOT_FOUND" in str(e):
            raise HTTPException(404, detail={"error": {"code": "PROJECT_NOT_FOUND"}})
        raise
    return {"success": True, "data": _read(new), "error": None}

File: web/src/api/projects.ts
import { api, ApiError } from "./client";

export type ProjectMode = "forward" | "reverse";
export type ProjectType = "dev_only" | "ops_only" | "dev_and_ops";
export type AssessmentKind = "development" | "enhancement";
export type ProjectPhase =
  | "budget"
  | "bidding"
  | "planning"
  | "change"
  | "settled";

export interface Project {
  id: string;
  name: string;
  project_type: ProjectType;
  assessment_kind: AssessmentKind;
  mode: ProjectMode;
  city: string;
  industry: string;
  phase: ProjectPhase;
  basis_data_ver: string;
  client?: string;
  evaluator?: string;
  target_cost?: number;
  other_cost?: number;
  include_ops?: boolean;
  alpha_dev?: number;
  fp_method?: "nesma_estimated" | "ifpug" | "quick";
  total_fp?: number;
  total_cost?: number;
  created_at: string;
  updated_at: string;
}

// v2.0 T8 — query/copy go through the new {success,data,meta} envelope.
// v2.0 T20 — list() now delegates to query() so both speak the new envelope.
// Callers that only care about the array (e.g. store.fetchAll) keep their
// existing Promise<Project[]> contract; callers that need meta/filter/sort
// (e.g. ProjectList toolbar) call query() directly.
export interface ProjectQuery {
  q?: string;
  city?: string;
  industry?: string;
  phase?: ProjectPhase;
  mode?: ProjectMode;
  sort?: "created_at" | "updated_at" | "name" | "target_cost";
  order?: "asc" | "desc";
  page?: number;
  size?: number;
}

export interface ProjectQueryResult {
  data: Project[];
  meta: { total: number; page: number; size: number };
}

// v2.7 — 项目批量导出 / 导入 bundle 类型。
export interface ProjectBundleItem {
  name: string;
  project_type: ProjectType;
  assessment_kind?: AssessmentKind;
  phase: ProjectPhase;
  city: string;
  industry: string;
  client?: string | null;
  evaluator?: string | null;
  mode: ProjectMode;
  target_cost?: number | null;
  other_cost?: number;
  include_ops?: boolean;
  alpha_dev?: number;
  fp_method?: "nesma_estimated" | "ifpug" | "quick";
  basis_data_ver: string;
  factors_dev?: Record<string, unknown> | null;
  factors_ops?: Record<string, unknown> | null;
  param_overrides: Array<{ key: string; value: string; reason?: string | null }>;
  function_points: Array<Record<string, unknown>>;
}

export interface ProjectBundle {
  version: string;
  exported_at: string;
  projects: ProjectBundleItem[];
}

export interface ProjectImportResult {
  imported: number;
  project_ids: string[];
}

interface NewEnvelope<T> {
  success: boolean;
  data: T;
  error: { code: string; message?: string; details?: Record<string, unknown> } | null;
  meta?: { total: number; page: number; size: number };
}

function unwrapNew<T>(payload: unknown): T {
  if (!payload || typeof payload !== "object") {
    throw new ApiError("INVALID_RESPONSE", "Server returned malformed envelope");
  }
  const env = payload as NewEnvelope<T>;
  if (env.success) return env.data;
  const e = env.error ?? { code: "UNKNOWN", message: "" };
  throw new ApiError(e.code, e.message ?? "", e.details);
}

function buildQueryString(opts: ProjectQuery): string {
  const params = new URLSearchParams();
  // Filter out undefined, null, and empty-string values — the backend treats
  // missing params as "no filter" and we don't want empty filters polluting
  // the URL.
  (Object.keys(opts) as Array<keyof ProjectQuery>).forEach((key) => {
    const value = opts[key];
    if (value !== undefined && value !== null && value !== "") {
      params.set(key, String(value));
    }
  });
  return params.toString();
}

export const projectsApi = {
  // list() returns just the array for backwards compatibility with the
  // projects store; full meta/filter/sort go through query().
  async list(): Promise<Project[]> {
    const { data } = await projectsApi.query();
    return data;
  },
  get: (id: string) => api.get<Project>(`/api/projects/${id}`),
  create: (body: Partial<Project>) => api.post<Project>("/api/projects", body),
  update: (id: string, body: Partial<Project>) =>
    api.patch<Project>(`/api/projects/${id}`, body),
  patch: (id: string, body: Partial<Project>) =>
    api.patch<Project>(`/api/projects/${id}`, body),
  remove: (id: string) => api.delete<void>(`/api/projects/${id}`),

  async query(opts: ProjectQuery = {}): Promise<ProjectQueryResult> {
    const qs = buildQueryString(opts);
    const url = `/api/projects${qs ? "?" + qs : ""}`;
    const resp = await api.raw.get<NewEnvelope<Project[]>>(url);
    const data = unwrapNew<Project[]>(resp.data);
    const meta = (resp.data as NewEnvelope<Project[]>).meta ?? {
      total: data.length,
      page: 1,
      size: data.length,
    };
    return { data, meta };
  },

  async copy(srcId: string, name: string): Promise<Project> {
    const resp = await api.raw.post<NewEnvelope<Project>>(
      `/api/projects/${srcId}/copy`,
      { name },
    );
    return unwrapNew<Project>(resp.data);
  },

  async exportProjects(ids: string[]): Promise<ProjectBundle> {
    const resp = await api.raw.post<NewEnvelope<ProjectBundle>>(
      "/api/projects/export",
      { ids },
    );
    return unwrapNew<ProjectBundle>(resp.data);
  },

  async importProjects(bundle: ProjectBundle): Promise<ProjectImportResult> {
    const resp = await api.raw.post<NewEnvelope<ProjectImportResult>>(
      "/api/projects/import",
      bundle,
    );
    return unwrapNew<ProjectImportResult>(resp.data);
  },
};
